test/integration/clustercontroller_test.go

```diff
@@ -0,0 +1,263 @@
+						Size: len(name), // using name length here just to give some variance to the node group size
+					},
+				}
+			}
+
+			clusterOperatorClient.ClusteroperatorV1alpha1().Clusters(testNamespace).Create(cluster)
+
+			if err := waitForClusterToExist(clusterOperatorClient, testNamespace, testClusterName); err != nil {
+				t.Fatalf("error waiting from Cluster to exist: %v", err)
+			}
+
+			if err := waitForClusterToReconcile(clusterOperatorClient, testNamespace, testClusterName); err != nil {
+				t.Fatalf("error waiting for Cluster to reconcile: %v", err)
+			}
+
+			nodeGroups, err := clusterOperatorClient.ClusteroperatorV1alpha1().NodeGroups(testNamespace).List(metav1.ListOptions{})
+			if err != nil {
+				t.Fatalf("error getting the node groups: %v", err)
+			}
+
+			if e, a := 1+len(tc.computeNodeGroupNames), len(nodeGroups.Items); e != a {
+				t.Fatalf("unexpected number of node groups: expected %v, got %v", e, a)
+			}
+
+			type details struct {
+				namePrefix string
+				nodeType   v1alpha1.NodeType
+				size       int
+			}
+
+			expectedDetails := make([]details, 1+len(tc.computeNodeGroupNames))
+			expectedDetails[0] = details{
+				namePrefix: fmt.Sprintf("%s-master", testClusterName),
+				nodeType:   v1alpha1.NodeTypeMaster,
+				size:       sizeOfMasterNodeGroup,
+			}
+			for i, name := range tc.computeNodeGroupNames {
+				expectedDetails[1+i] = details{
+					namePrefix: fmt.Sprintf("%s-compute-%s", testClusterName, name),
+					nodeType:   v1alpha1.NodeTypeCompute,
+					size:       len(name),
+				}
+			}
+
+			actualDetails := make([]details, len(nodeGroups.Items))
+			for i, ng := range nodeGroups.Items {
+				lastDashInName := strings.LastIndexByte(ng.Name, '-')
+				if lastDashInName < 0 {
+					t.Fatalf("NodeGroup name does not contain any dashes")
+				}
+				actualDetails[i] = details{
+					namePrefix: ng.Name[:lastDashInName],
+					nodeType:   ng.Spec.NodeType,
+					size:       ng.Spec.Size,
+				}
+			}
+
+			sortDetails := func(detailsSlice []details) {
+				sort.Slice(detailsSlice, func(i, j int) bool {
+					return detailsSlice[i].namePrefix < detailsSlice[j].namePrefix
+				})
+			}
+			sortDetails(expectedDetails)
+			sortDetails(actualDetails)
+
+			if e, a := len(expectedDetails), len(actualDetails); e != a {
+				t.Fatalf("unexpected number of node groups: expected %q, got %q", e, a)
+			}
+
+			for i, a := range actualDetails {
+				e := expectedDetails[i]
+				if e != a {
+					t.Fatalf("unexpected node group: expected %+v, got %+v", e, a)
+				}
+			}
+		})
+	}
+}
+
+// startServerAndClusterController creates a new test ClusterController injected with
+// fake clients and returns:
+//
+// - a fake kubernetes core api client
+// - a cluster-operator api client
+// - a function for shutting down the controller and API server
+//
+// If there is an error, startServerAndClusterController calls 'Fatal' on the
+// injected testing.T.
+func startServerAndClusterController(t *testing.T) (
```

Comment:
I really like this approach, it's a big win IMO to be able to just bring up our apiserver and etcd independently. However when we get to our controllers, can we integration test those in this same manner? The infra controller for example will attempt to create configmaps and jobs. We could stub that such that the pods it will launch mock what they would do somehow, but without a Kube API server + controllers, am I correct that that aspect of our code can't be integration tested in this fashion? 

We can mimic an awful lot of that with straight unit tests however, so maybe that's good enough, but I'm wondering what you envision for our controllers running ansible in this integration suite.

Reply:
You can have all the controllers, too, in your integration tests.  See 
https://github.com/openshift/online/blob/master/accountant/test/integration/helper.go#L103

There's `StartTestMaster` and there's `StartTestMasterAPI`.  The former has the API and all controllers (but not kubelet). The latter is just API.

Reply:
Yeah that should do it, but I think it means vendoring a whole lot more code. (all of origin)

Reply:
The interactions with the Kube API server is done via the fake kube client. The integration tests would verify that the expected actions were taken and would establish reactions to those actions. This is the same as what we should expect to do with the AWS fake. The Kube API server is outside of the responsibility of what the cluster-operator integration tests should be testing.

Reply:
Oh, I see the problem. The issue is that the infra controller is creating a pod to do the work, rather than doing it in-process. I don't think that this integration test framework is going to be able to test the behavior inside the pod, as it is.

Reply:
It's a question of where we draw the line of integration. If we could hit a point where we could literally run ansible pods and hit that mock AWS thing @twiest mentioned the other day, that would be about perfect.

If not, we could rig it such that it will launch jobs that just no-op and report success, and update status accordingly to unblock the other phases of the provisioning.

Or we can not launch the jobs at all. We can get an awful lot of coverage on the controllers with unit tests.